CRM/Upgrade/Incremental/php/FiveTwentyEight.php

```diff
@@ -40,10 +46,16 @@ public function setPreUpgradeMessage(&$preUpgradeMessage, $rev, $currentVer = NU
    *   an intermediate version; note that setPostUpgradeMessage is called repeatedly with different $revs.
    */
   public function setPostUpgradeMessage(&$postUpgradeMessage, $rev) {
-    // Example: Generate a post-upgrade message.
-    // if ($rev == '5.12.34') {
-    //   $postUpgradeMessage .= '<br /><br />' . ts("By default, CiviCRM now disables the ability to import directly from SQL. To use this feature, you must explicitly grant permission 'import SQL datasource'.");
-    // }
+    // Example: Generate a pre-upgrade message.
+    if (!function_exists('civi_wp')) {
+      //exit
+    }
+    elseif ($rev == '5.28.alpha1') {
+      $postUpgradeMessage .= '<p>' . ts('Starting with version 5.29.0, CiviCRM on WP will
```

Comment:
Looks like a copy and paste from above if above is accepted.

Reply:
any upgrade that 'passes through' 5.28 will get this so 4.7.31 to 5.28 will get it as well as 5.26 to 5.29.   We might also want it in 5.29 but I have to ask how we handle that kind of notice.

Reply:
Oh, yeah. That's a funny edge-case. 🤔 

So, on the copy-paste for the pre-upgrade/post-upgrade message... you might put the messaging into a separate function. This makes it easier to reuse/rejigger. eg

```php
  if ($rev == '5.28.alpha1') {
     $preUpgradeMessage .= CRM_Upgrade_Incremental_php_FiveTwentyEight::createWpFilesMessage();
  }

    ...

  public static function createWpFilesMessage() {
    if (!function_exists('civi_wp')) {
      return '';
    }
    else {
      return '<p>' . ts('...The messaging...') . '</p>';
    }
  }
```

If I read correctly, I think you're suggesting that we want this behavior:

* `5.25` => `5.28` - Show message once pre-upgrade and once post-upgrade
* `5.28` => `5.29` - Show message once pre-upgrade and once post-upgrade
* `5.25` => `5.29` - Show message once pre-upgrade and once post-upgrade

I think it could be done by putting *different code in each branch* (5.28 vs master). Ex:

* For this PR for `5.28`, it patches `FiveTwentyEight.php`. Both `setPreUpgradeMessage()` and `setPostUpgradeMessage()` call `createWpFilesMessage()`.
* After that's merged into `5.28`, the changes propagate to `master`.
* Add another PR for `master`/`5.29`. *Remove* the calls from `FiveTwentyEight.php` and add the calls to `FiveTwentyNine.php`.

This is a bit messy on the git side, but it keeps the code simple/traceable, and I'm hard-pressed to find something else that hits the described behavior in all upgrader UIs.

Aside: It might make sense to target the message even more precisely. On my local, this distinguished between old-style (4.6) and new-style builds:

```php
  public static function createWpFilesMessage() {
    if (!function_exists('wp_get_upload_dir')) {
      // Not a WordPress build
      return '';
    }

    if (realpath(wp_get_upload_dir()['basedir'] . '/civicrm') === realpath(CRM_Utils_File::baseFilePath())) {
      // Old and new formulas lead to the same place. 
      return '';
    }
    
    if (isset($GLOBALS['civicrm_paths']['civicrm.files']['path'])) {
      // They've explicitly chosen to use a non-default path.
      return '';
    }

    return '<p>' . ts('...The messaging...') . '</p>';
  }
```



The main consideration is duplicate messaging - e.g. if someone upgrades from 5.25=>5.29, you don't want them to get the message 





(There's other trickery one might work )

The main issue with reproducing the message in 5.28 and 5.29 is actually with *duplicate* messages.